### PR TITLE
Bindings for the new extension system

### PR DESCRIPTION
**Pull request by vnen**
**This is still not 100% functional and needs lots of testing.**

This recreates the whole repository to work with the extension system that replaces GDNative in Godot 4.0.

The API is changed to be close to the one used in Godot itself to diminish the differences between an extension and a module (technically you could use the same code for both using a few `#ifdef`s).

### Missing things

- CI checks, including a proper test script. The workflow file is still the same but it isn't backed by the repo scripts.
- Moving math types to inlined structures. For now it uses generated wrappers, without direct member access.
   - Note that the new API JSON file includes sizes and offsets for the builtin types so there should be validation when the inlined structs are done.
- The ClassDB singleton from Godot conflicts with the one from this repository, so for now the engine one isn't available (being skipped during generation). They should be merged somehow (I do have some ideas how to do it but nothing simple so haven't done it yet).
- Linking to `godot-headers` as a submodule. That repo still needs an update for `master` so for now this includes the needed files in `godot-headers-temp`.
- Documentation.
- Probably many more things that I'm forgetting now.

### How it works

The README file has a laconic description of the process. Check the demo project in the `test` folder to have an idea.

You can regenerate the API JSON from Godot by running `godot --dump-extension-api`. The file `extension_api.json` will be created in the same folder from which you run the command.

### CODE REVIEW

**Comment by bruvzg** [comment]
Some issues encountered during https://github.com/godotengine/godot/pull/52192 development:

* Casting references fails:
``` c++
	Ref<TextServerTest> ts;
	ts.instantiate();
	/* void add_interface(const Ref<TextServer> &p_interface); */
	TextServerManager::get_singleton()->add_interface(ts);
```
```
src/register_types.cpp:19:54: error: no viable conversion from 'Ref<godot::TextServerTest>' to 'const Ref<godot::TextServer>'
                        TextServerManager::get_singleton()->add_interface(ts);
                                                                          ^~
godot-cpp/include/godot_cpp/classes/ref.hpp:167:2: note: candidate constructor not viable: no known conversion from 'Ref<godot::TextServerTest>' to 'const godot::Ref<godot::TextServer> &' for 1st argument
        Ref(const Ref &p_from) {
        ^
godot-cpp/include/godot_cpp/classes/ref.hpp:171:2: note: candidate constructor not viable: no known conversion from 'Ref<godot::TextServerTest>' to 'godot::TextServer *' for 1st argument
        Ref(T *p_reference) {
        ^
godot-cpp/include/godot_cpp/classes/ref.hpp:177:2: note: candidate constructor not viable: no known conversion from 'Ref<godot::TextServerTest>' to 'const godot::Variant &' for 1st argument
        Ref(const Variant &p_variant) {
        ^
godot-cpp/include/godot_cpp/classes/ref.hpp:113:2: note: candidate function
        operator Variant() const {
```

* Virtual methods with `enum` arguments fails:

```c++
class TextServerTest : public TextServerExtension {
	GDCLASS(TextServerTest, TextServerExtension);
....
	virtual bool _has_feature(TextServer::Feature feature) override;
```
```
In file included from godot-cpp/include/godot_cpp/core/class_db.hpp:38:
In file included from godot-cpp/include/godot_cpp/core/method_bind.hpp:34:
godot-cpp/include/godot_cpp/core/binder_common.hpp:162:59: error: no member named 'convert' in 'godot::PtrToArg<godot::TextServer::Feature>'
        PtrToArg<R>::encode((p_instance->*p_method)(PtrToArg<P>::convert(p_args[Is])...), r_ret);
```

* Virtual methods with `GDNativePtr<>` arguments fails:
```c++
class TextServerTest : public TextServerExtension {
	GDCLASS(TextServerTest, TextServerExtension);
....
	virtual void _shaped_text_get_glyphs(const RID &shaped, Glyph *gl_array) const override;
```
```
In file included from godot-cpp/include/godot_cpp/core/class_db.hpp:38:
In file included from godot-cpp/include/godot_cpp/core/method_bind.hpp:34:
In file included from godot-cppinclude/godot_cpp/core/binder_common.hpp:36:
godot-cpp/include/godot_cpp/core/method_ptrcall.hpp:168:122: error: no member named '___binding_callbacks' in 'godot::Glyph'
                return reinterpret_cast<T *>(godot::internal::interface->object_get_instance_binding(p_ptr, godot::internal::token, T::___binding_callbacks));
```

**Comment by Shatur** [comment]
I faced the same problems as @bruvzg. But I also want to add that virtual functions overloads such as `_ready()` are not called at all.

**Comment by bruvzg** [comment]
And another inconvenience with the current implementation, there's no clear way to add  a custom callbacks for init levels. GDExtensionBinding should have it's (de)initialize methods virtual. And we probably should have extra init levels after singleton registration, or change the init order on the engine side.

**Comment by Shatur** [comment]
> But I also want to add that virtual functions overloads such as _ready() are not called at all.

Double checked, the issue related to the `_ready()`. Other virtual functions, like `_has_point` works as expected.

**Comment by BastiaanOlij** [comment]
@bruvzg so far for me just having `Ref<Class>` fails, haven't gotten much further with that one though, interesting that you're able to define it but are running into trouble using it.

**Comment by bruvzg** [comment]
> interesting that you're able to define it but are running into trouble using it.

It's building, but `instantiate` call crashes, `memnew` also crashes.

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,75 +1,65 @@
...
 # Todo
 # Test build for Windows, Mac and mingw.
 
-project(godot-cpp)
+project(godot-cpp LANGUAGES CXX)
 cmake_minimum_required(VERSION 3.6)
 
 option(GENERATE_TEMPLATE_GET_NODE "Generate a template version of the Node class's get_node." ON)
 
-# Change the output directory to the bin directory
```
Comment (BastiaanOlij) [inline CMakeLists.txt:42]: these may have ended up being removed by accident...

Comment (Shatur) [reply CMakeLists.txt:42]: No, it was on purpose

Comment (Shatur) [reply CMakeLists.txt:42]: Here is my PR: https://github.com/vnen/godot-cpp/pull/14

Comment (BastiaanOlij) [reply CMakeLists.txt:42]: All good, one less thing to worry about :)
It's really hard to see what changes were part of our rewrite, and which are undoing PRs that were applied on master after they diverged.

Comment (BastiaanOlij) [reply CMakeLists.txt:42]: And in the end it's probably going to mean we're going to need to re-evaluate some fixes.

**Review by akien-mga — APPROVED** [review]
Commit history somewhat messy and there's too many changes for me to do an in-depth code review, but it's been tested and co-developed by many contributors so it should be good to merge now and refine in `master`.

**Pull request merged**